### PR TITLE
[Merged by Bors] - Add missing `ReadOnly = Self` bound

### PR DESCRIPTION
**Pull request by BoxyUwU**
# Objective
`ReadOnlyWorldQuery` should have required `Self::ReadOnly = Self` so that calling `.iter()` on a readonly query is equivelent to calling `iter_mut()`.

## Solution

add `ReadOnly = Self` to the definition of `ReadOnlyWorldQuery`

---

## Changelog

ReadOnlyWorldQuery's `ReadOnly` assoc type is now always equal to `Self`

## Migration Guide

Make `Self::ReadOnly = Self` hold

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]

**Comment by nicopap** [comment]
I can't review this. I've very little understanding of the implications. Though this was triggered by some code samples I posted on discord.

**Comment by alice-i-cecile** [comment]
To explain this PR in plain English:

1. All types that implement the `WorldQuery` trait have a `ReadOnly` associated type, which describe the type that the query becomes if it's turned into its read-only variant.
2. Queries which are read-only also implement the `ReadOnlyWorldQuery` trait, which is a subtrait of the `WorldQuery` trait as all read-only world queries are also ordinary world queries.
3. When implementing the `ReadOnlyWorldQuery` trait, we must guarantee that turning a read-only query into it's read-only equivalent (bullet point 1), nothing happens. This is what the `WorldQuery<ReadOnly = Self>` bound means: this ensures that this transformation is always an identity function.
4. This both improves correctness and makes type inference play nicer.

Ping @harudagondi for review :)

**Review by nicopap — APPROVED** [review]

**Review by IceSentry — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
bors retry

**Comment by alice-i-cecile** [comment]
bors retry

**Review by TheRawMeatball — APPROVED** [review]
